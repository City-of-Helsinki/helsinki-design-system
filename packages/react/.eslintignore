**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
