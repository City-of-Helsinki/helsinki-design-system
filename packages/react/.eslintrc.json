{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
