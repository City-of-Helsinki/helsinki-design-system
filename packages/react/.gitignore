node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
