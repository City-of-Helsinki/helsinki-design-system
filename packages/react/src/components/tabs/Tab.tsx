import React, { useContext, useEffect, useRef } from 'react';

// import core base styles
import 'hds-core';

import styles from './Tabs.module.scss';
import classNames from '../../utils/classNames';
import { TabsContext } from './TabsContext';

export type TabProps = React.PropsWithChildren<{
  /**
   * Custom function callback for on tab click
   */
  onClick?: () => void;
  /**
   * Additional class names to apply to the Tab
   */
  className?: string;
  /**
   * Additional styles to apply to the Tab
   */
  style?: React.CSSProperties;
  index?: number;
}>;

export const Tab = ({ children, className, index, onClick, style }: TabProps) => {
  const ref = useRef<HTMLLIElement>(null);
  const { activeTab, focusedTab, setFocusedTab, setActiveTab } = useContext(TabsContext);
  const isActive = activeTab === index;

  /**
   * Focus the tab based on focusedTab provided by the context
   */
  useEffect(() => {
    if (focusedTab === index) {
      ref.current.focus({ preventScroll: true });
    }
  }, [focusedTab, index]);

  /**
   * Set the tab as active when clicked
   */
  const onTabClick = () => {
    if (onClick) {
      onClick();
    }
    setActiveTab(index);
    setFocusedTab(index);
  };

  /**
   * Set the tab as active when enter is pressed
   * @param event
   */
  const onKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
    const isEnter = event.key === 'Enter' || event.keyCode === 13;
    const isSpace = event.key === ' ' || event.keyCode === 32;
    if (isEnter || isSpace) {
      setActiveTab(index);
    }
  };

  /**
   * Handle the focus event
   */
  const onFocus = (event: React.FocusEvent<HTMLLIElement>) => {
    event.preventDefault();
    if (focusedTab !== index) {
      setFocusedTab(index);
    }
  };

  return (
    <li
      role="tab"
      aria-selected={isActive}
      aria-controls={`tab-${index}-panel`}
      tabIndex={isActive ? 0 : -1}
      id={`tab-${index}-button`}
      className={classNames(styles.tab, isActive && styles.active, className)}
      style={style}
      onClick={onTabClick}
      onKeyDown={onKeyDown}
      onFocus={onFocus}
      ref={ref}
    >
      <span>{children}</span>
    </li>
  );
};
Tab.componentName = 'Tab';
