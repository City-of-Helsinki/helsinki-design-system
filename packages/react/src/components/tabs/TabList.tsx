import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';

import '../../styles/base.module.css';
import styles from './Tabs.module.scss';
import classNames from '../../utils/classNames';
import { FCWithName } from '../../common/types';
import { TabsContext } from './TabsContext';
import { IconAngleLeft, IconAngleRight } from '../../icons';
import { getChildByIndex, isElementOutsideLeftEdge, isElementOutsideRightEdge } from './tabUtils';
import { getChildrenAsArray } from '../../utils/getChildren';

export type TabListProps = React.PropsWithChildren<{
  /**
   * Additional class names to apply to the Tablist
   */
  className?: string;
  /**
   * Additional styles to apply to the Tablist
   */
  style?: React.CSSProperties;
}>;

export const TabList = ({ children, className, style = {} }: TabListProps) => {
  const tablistElementRef = useRef<HTMLUListElement>(null);
  const tablistContainerRef = useRef<HTMLDivElement>(null);
  const { focusedTab, setFocusedTab } = useContext(TabsContext);
  const [tablistWidth, setTablistWidth] = useState<number>(0);
  const [scrollIndex, setScrollIndex] = useState<number>(null);
  const [showPreviousButton, setShowPreviousButton] = useState(true);
  const [showNextButton, setShowNextButton] = useState(true);
  const [scrollValue, setScrollValue] = useState<number>(0);

  const childElements = getChildrenAsArray(children);

  /**
   * Pass the index as prop to each tab element
   */
  const tabs = childElements.map((child, index) => {
    if (React.isValidElement(child) && (child.type as FCWithName).componentName === 'Tab') {
      // eslint-disable-next-line react/no-array-index-key
      return React.cloneElement(child, { index, key: index });
    }
    return null;
  });

  /**
   * Calculate the scroll position
   */
  const updateScrollPosition = useCallback(() => {
    // Get the tab element by the index
    const tabElement = getChildByIndex(scrollIndex, tablistElementRef.current);
    if (tabElement !== null) {
      // Calculate the distance of tab edges to the left edge of the tablist
      const tabRightEdgeDistance = tabElement.offsetLeft + tabElement.offsetWidth;
      const tabLeftEdgeDistance = tabElement.offsetLeft;
      // Calculate how much tab right edge is outside the tablist container
      const tabListContainerWidth = tablistContainerRef.current.offsetWidth;
      const tabRightEdgeOffset = tabRightEdgeDistance - tabListContainerWidth;
      // Check if tab right edge is outside the right edge of the container
      if (
        tabRightEdgeOffset > 0 &&
        (scrollValue < tabRightEdgeOffset || scrollIndex === tablistElementRef.current.children.length - 1)
      ) {
        setScrollValue(tabRightEdgeOffset);
      }
      // Check if tab left edge is outside the left edge of the container
      else if (tabLeftEdgeDistance < scrollValue) {
        setScrollValue(tabLeftEdgeDistance);
      }
    }
  }, [scrollIndex, scrollValue]);

  /**
   * Find the next tab that is outside the right edge of the tablist container
   */
  const findNextElementOutsideRightEdge = useCallback(() => {
    const tabListElement = tablistElementRef.current;
    const tabIndex = Array.from(tabListElement.children).findIndex((element) =>
      isElementOutsideRightEdge(element, tablistContainerRef.current, scrollValue),
    );
    return tabIndex;
  }, [scrollValue]);

  /**
   * Find the next tab that is outside the left edge of the tablist
   */
  const findNextElementOutsideLeftEdge = useCallback(() => {
    const tabListElement = tablistElementRef.current;
    const tabIndex = Array.from(tabListElement.children)
      .reverse()
      .findIndex((element) => isElementOutsideLeftEdge(element, scrollValue));
    return tabIndex > -1 ? tabListElement.children.length - tabIndex - 1 : -1;
  }, [scrollValue]);

  /**
   * Hide/show scroll buttons
   */
  const updateScrollButtons = useCallback(() => {
    setShowPreviousButton(findNextElementOutsideLeftEdge() !== -1);
    setShowNextButton(findNextElementOutsideRightEdge() !== -1);
  }, [findNextElementOutsideLeftEdge, findNextElementOutsideRightEdge]);

  /**
   * Calculate tablist width
   */
  useEffect(() => {
    const totalWidth = Array.from(tablistElementRef.current.children).reduce((total, el) => {
      return total + (el as HTMLElement).offsetWidth;
    }, 0);
    setTablistWidth(totalWidth);
  }, [children]);

  /**
   * Update scroll position when focus changes
   */
  useEffect(() => {
    setScrollIndex(focusedTab);
  }, [focusedTab]);

  /**
   * Hide/show next/prev buttons when necessary
   */
  useEffect(() => {
    updateScrollButtons();
    window.addEventListener('resize', updateScrollButtons);
    return () => {
      window.removeEventListener('resize', updateScrollButtons);
    };
  }, [updateScrollButtons]);

  /**
   * Update the scroll position when buttons are hidden/shown
   */
  useEffect(() => {
    updateScrollPosition();
  }, [updateScrollPosition, scrollIndex, showPreviousButton, showNextButton]);

  /**
   * Handle the next button click
   */
  const onNextButtonClick = () => {
    const nextTabIndex = findNextElementOutsideRightEdge();
    if (nextTabIndex > -1) {
      setScrollIndex(nextTabIndex);
    }
  };

  /**
   * Handle the previous button click
   */
  const onPreviousButtonClick = () => {
    const nextTabIndex = findNextElementOutsideLeftEdge();
    if (nextTabIndex > -1) {
      setScrollIndex(nextTabIndex);
    }
  };

  /**
   * Handle keyboard arrow keys
   * @param event
   */
  const onKeyDown = (event: React.KeyboardEvent<HTMLUListElement>) => {
    let nextIndex: number = null;
    if (event.key === 'ArrowRight') {
      nextIndex = Math.min(focusedTab + 1, tabs.length - 1);
    }
    if (event.key === 'ArrowLeft') {
      nextIndex = Math.max(focusedTab - 1, 0);
    }
    if (nextIndex !== null) {
      setFocusedTab(nextIndex);
    }
  };

  return (
    <div className={classNames(styles.tablistBar, className)} style={style}>
      {showPreviousButton && (
        <div className={styles.scrollButton} aria-hidden="true">
          <button type="button" onClick={onPreviousButtonClick} tabIndex={-1}>
            <IconAngleLeft />
          </button>
        </div>
      )}
      <div className={styles.tablist} ref={tablistContainerRef}>
        <ul
          role="tablist"
          ref={tablistElementRef}
          onKeyDown={onKeyDown}
          style={{
            width: `${tablistWidth}px`,
            transform: scrollValue ? `translateX(${-1 * scrollValue}px)` : undefined,
          }}
        >
          {tabs}
        </ul>
      </div>
      {showNextButton && (
        <div className={styles.scrollButton} aria-hidden="true">
          <button type="button" onClick={onNextButtonClick} tabIndex={-1}>
            <IconAngleRight />
          </button>
        </div>
      )}
    </div>
  );
};
TabList.componentName = 'TabList';
