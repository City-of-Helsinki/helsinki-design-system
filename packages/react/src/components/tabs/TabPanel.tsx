import React, { useContext } from 'react';

// import core base styles
import 'hds-core';
import { TabsContext } from './TabsContext';

export type TabPanelProps = React.PropsWithChildren<{
  /**
   * Additional class names to apply to the Tab
   */
  className?: string;
  /**
   * Additional styles to apply to the Tab
   */
  style?: React.CSSProperties;
  index?: number;
}>;

export const TabPanel = ({ children, className, index, style }: TabPanelProps) => {
  const { activeTab } = useContext(TabsContext);
  const isActive = activeTab === index;

  return isActive ? (
    <div
      id={`tab-${index}-panel`}
      role="tabpanel"
      aria-labelledby={`tab-${index}-button`}
      className={className}
      style={style}
    >
      {children}
    </div>
  ) : null;
};
TabPanel.componentName = 'TabPanel';
