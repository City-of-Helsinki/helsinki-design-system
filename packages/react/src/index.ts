export * from './icons';
export * from './components';
export * from './ssr';
