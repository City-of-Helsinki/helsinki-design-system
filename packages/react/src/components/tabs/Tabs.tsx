import React, { useState } from 'react';

// import core base styles
import 'hds-core';
import styles from './Tabs.module.scss';
import classNames from '../../utils/classNames';
import { TabsContext } from './TabsContext';
import { FCWithName } from '../../common/types';
import { useTheme } from '../../hooks/useTheme';
import { TabList } from './TabList';
import { TabPanel } from './TabPanel';
import { Tab } from './Tab';

export interface TabsCustomTheme {
  '--tablist-border-color'?: string;
  '--tablist-border-size'?: string;
  '--tab-color'?: string;
  '--tab-font-size'?: string;
  '--tab-height'?: string;
  '--tab-min-width'?: string;
  '--tab-spacing'?: string;
  '--tab-active-border-size'?: string;
  '--tab-active-border-color'?: string;
  '--tab-focus-outline-size'?: string;
  '--tab-focus-outline-color'?: string;
}

export type TabsProps = React.PropsWithChildren<{
  /**
   * The initially active tab
   */
  initiallyActiveTab?: number;
  /**
   * Use the small variant
   * @default false
   */
  small?: boolean;
  /**
   * Defines the tabs theme
   */
  theme?: TabsCustomTheme;
}>;

export const Tabs = ({ children, initiallyActiveTab = 0, small = false, theme }: TabsProps) => {
  const [activeTab, setActiveTab] = useState<number>(initiallyActiveTab);
  const [focusedTab, setFocusedTab] = useState<number>(null);
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<TabsCustomTheme>(styles.tabs, theme);

  /**
   * Get the TabList from children
   */
  const tabList = React.Children.toArray(children).filter((child) => {
    return React.isValidElement(child) && (child.type as FCWithName).componentName === 'TabList';
  });

  /**
   * Get TabPanels from children
   */
  const tabPanels = React.Children.toArray(children)
    .filter((child) => {
      return React.isValidElement(child) && (child.type as FCWithName).componentName === 'TabPanel';
    })
    .map((child, index) => {
      if (React.isValidElement(child)) {
        // Pass index prop to the TabPanel
        return React.cloneElement(child, { index });
      }
      return child;
    });

  return (
    <TabsContext.Provider value={{ activeTab, setActiveTab, focusedTab, setFocusedTab }}>
      <div className={classNames(styles.tabs, small && styles.small, theme && customThemeClass)}>
        {tabList}
        {tabPanels}
      </div>
    </TabsContext.Provider>
  );
};

// Using the Tabs component and its child components in the same namespace ensures that the group is considered as a single component.
// Tabs components do some child tab component processing and this might break in some preprocessed environments if the components were separate components, for example in mdx environment.
Tabs.TabList = TabList;
Tabs.TabPanel = TabPanel;
Tabs.Tab = Tab;
