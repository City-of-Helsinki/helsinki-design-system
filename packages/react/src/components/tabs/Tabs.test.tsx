import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { Tabs } from './Tabs';
import { TabList } from './TabList';
import { Tab } from './Tab';
import { TabPanel } from './TabPanel';

describe('<Tabs /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(
      <Tabs>
        <TabList>
          <Tab>Foo</Tab>
          <Tab>Bar</Tab>
        </TabList>
        <TabPanel>Fizz</TabPanel>
        <TabPanel>Buzz</TabPanel>
      </Tabs>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(
      <Tabs>
        <TabList>
          <Tab>Foo</Tab>
          <Tab>Bar</Tab>
        </TabList>
        <TabPanel>Fizz</TabPanel>
        <TabPanel>Buzz</TabPanel>
      </Tabs>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should call onClick when user clicks tab', () => {
    const onClick = jest.fn();
    render(
      <Tabs>
        <TabList>
          <Tab onClick={onClick}>Foo</Tab>
          <Tab onClick={onClick}>Bar</Tab>
        </TabList>
        <TabPanel>Fizz</TabPanel>
        <TabPanel>Buzz</TabPanel>
      </Tabs>,
    );
    userEvent.click(screen.getByLabelText('Foo'));
    waitFor(() => expect(onClick).toHaveBeenCalled());
  });

  it('should call onClick when user presses enter on tab', () => {
    const onClick = jest.fn();
    render(
      <Tabs>
        <TabList>
          <Tab onClick={onClick}>Foo</Tab>
          <Tab onClick={onClick}>Bar</Tab>
        </TabList>
        <TabPanel>Fizz</TabPanel>
        <TabPanel>Buzz</TabPanel>
      </Tabs>,
    );
    userEvent.type(screen.getByLabelText('Foo'), '{Enter}');
    waitFor(() => expect(onClick).toHaveBeenCalled());
  });
});
