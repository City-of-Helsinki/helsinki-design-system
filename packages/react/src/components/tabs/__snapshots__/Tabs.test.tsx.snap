// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="tabs"
  >
    <div
      class="tablistBar"
    >
      <div
        class="tablist"
      >
        <ul
          role="tablist"
          style="width: 0px;"
        >
          <li
            aria-controls="tab-0-panel"
            aria-selected="true"
            class="tab active"
            id="tab-0-button"
            role="tab"
            tabindex="0"
          >
            <span>
              Foo
            </span>
          </li>
          <li
            aria-controls="tab-1-panel"
            aria-selected="false"
            class="tab"
            id="tab-1-button"
            role="tab"
            tabindex="-1"
          >
            <span>
              Bar
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-labelledby="tab-0-button"
      id="tab-0-panel"
      role="tabpanel"
    >
      Fizz
    </div>
  </div>
</DocumentFragment>
`;
